### PR TITLE
Parquet: Improve Test Coverage of RowGroupFilter Code with Nans #6518

### PR DESCRIPTION
**Pull request by youngxinler**
This PR:
1. Add TestMetricsRowGroupFilter  "some_double_nans" field  which contain NaN as well as non-nan values for Coverage. 
2. Change TestMetricsRowGroupFilterTypes "double" and "float" fields , add some NAN, for value Coverage.

TestMetricsRowGroupFilterTypes. Regarding adding new fields containing NaN, or adding NAN to existing fields, I think the latter is better, because all field types in this class are unique.

#6518

### CODE REVIEW

**Comment by youngxinler** [comment]
@RussellSpitzer  can you help with the review?  About improve Test Coverage of RowGroupFilter Code with Nans.

File: data/src/test/java/org/apache/iceberg/data/TestMetricsRowGroupFilter.java [hunk 1]
```diff
@@ -341,6 +345,25 @@ public void testNotNaN() {
     Assert.assertTrue("Should read: NaN counts are not tracked in Parquet metrics", shouldRead);
   }
 
+
+  @Test
+  public void testDoubleWithNan() {
```
Comment (RussellSpitzer) [inline data/src/test/java/org/apache/iceberg/data/TestMetricsRowGroupFilter.java:350]: For this and the Float test we probably want some of the negative tests as well but they would only work in ORC. So I think just add a "if (orc) {

negative cases 
}

Comment (youngxinler) [reply data/src/test/java/org/apache/iceberg/data/TestMetricsRowGroupFilter.java:350]: thanks for @RussellSpitzer suggestion,  but when i run test as your suggestion.  
I found ORC also can't skip read.  so i think the negative test is work for parquet and ORC? 

as for #6517,  "// Only ORC should be able to distinguish using min/max when NaN is present",  i add these tests.
```java
      // just for "format == FileFormat.ORC"
      // record.setField("_some_double_nans", (i % 10 == 0) ? Double.NaN : 2D); // includes some nan values
      shouldRead = shouldRead(equal("some_double_nans", 10.0));
      Assert.assertFalse("Should skip: column with some nans contains target value", shouldRead);

      shouldRead = shouldRead(greaterThan("some_double_nans", 10.0));
      Assert.assertFalse("Should skip: column with some nans contains target value", shouldRead);
```
bacause the some_double_nans Max Value is 2D,  assert showRead will return false,  but all shouldRead is true,  it looks strange? 
If my guess is correct, the pushdown of ORC containing nans will not work. Is this as expected? Should we open a new issue to discuss?

image

**Comment by youngxinler** [comment]
@RussellSpitzer  Can I trouble you if you have time to do a review? About improve Test Coverage of RowGroupFilter Code with Nans. I have made changes.

File: data/src/test/java/org/apache/iceberg/data/TestMetricsRowGroupFilterTypes.java [hunk 2]
```diff
@@ -144,8 +144,8 @@ public void createInputFile() throws IOException {
       record.setField("_boolean", false);
       record.setField("_int", i);
       record.setField("_long", 5_000_000_000L + i);
-      record.setField("_float", ((float) (100 - i)) / 100F + 1.0F); // 2.0f, 1.99f, 1.98f, ...
-      record.setField("_double", ((double) i) / 100.0D + 2.0D); // 2.0d, 2.01d, 2.02d, ...
+      record.setField("_float", (i % 10 == 0) ? Float.NaN : ((float) (100 - i)) / 100F + 1.0F); // 2.0f, 1.99f, 1.98f, ... OR NAN
+      record.setField("_double", (i % 10 == 0) ? Double.NaN : ((double) i) / 100.0D + 2.0D); // 2.0d, 2.01d, 2.02d, ... OR NAN
```
Comment (RussellSpitzer) [inline data/src/test/java/org/apache/iceberg/data/TestMetricsRowGroupFilterTypes.java:148]: We probably shouldn't do that here. we can do negative tests as long as there aren't NANs correct?

File: data/src/test/java/org/apache/iceberg/data/TestMetricsRowGroupFilter.java [hunk 3]
```diff
@@ -320,6 +323,24 @@ public void testFloatWithNan() {
     Assert.assertTrue(shouldRead);
   }
 
+  @Test
+  public void testDoubleWithNan() {
```
Comment (RussellSpitzer) [inline data/src/test/java/org/apache/iceberg/data/TestMetricsRowGroupFilter.java:327]: +1, If ORC doesn't support it either this is ok with me

File: data/src/test/java/org/apache/iceberg/data/TestMetricsRowGroupFilterTypes.java (hunk 2 above)
Comment (RussellSpitzer) [reply data/src/test/java/org/apache/iceberg/data/TestMetricsRowGroupFilterTypes.java:148]: In testEq() it checks that the first parameterized value is read, and the second value is not read. So if there are NANs this should fail, if it doesn't fail with this code change let's figure out why.

**Comment by youngxinler** [comment]
please take a review if you have time. @RussellSpitzer

File: data/src/test/java/org/apache/iceberg/data/TestMetricsRowGroupFilter.java [hunk 4]
```diff
@@ -86,6 +65,27 @@
 import org.junit.runner.RunWith;
 import org.junit.runners.Parameterized;
 
+import static org.apache.iceberg.avro.AvroSchemaUtil.convert;
```
[the remark this answers is not in the document]
Comment (RussellSpitzer) [reply data/src/test/java/org/apache/iceberg/data/TestMetricsRowGroupFilter.java:68]: This breaks our checkstyle code, please run all relevant tests and check ```./gradlew build -x test  -x integrationTest``` before asking for review.

**Review by RussellSpitzer — APPROVED** [review]

**Comment by RussellSpitzer** [comment]
Thanks for the PR @youngxinler !

**Comment by youngxinler** [comment]
Thanks for the review and guidance in the process. @RussellSpitzer

**Pull request merged**